networktable.c

```diff
@@ -0,0 +1,337 @@
+				free(macVendor);
+		}
+		// Device in database AND client known to Pi-hole
+		else if(clientKnown)
+		{
+			// Update lastQuery. Only use new value if larger
+			// clients[clientID].lastQuery may be zero if this
+			// client is only known from a database entry but has
+			// not been seen since then
+			dbquery("UPDATE network "\
+			        "SET lastQuery = MAX(lastQuery, %ld) "\
+			        "WHERE id = %i;",\
+			        clients[clientID].lastQuery, dbID);
+
+			// Update numQueries. Add queries seen since last update
+			// and reset counter afterwards
+			dbquery("UPDATE network "\
+			        "SET numQueries = numQueries + %u "\
+			        "WHERE id = %i;",\
+			        clients[clientID].numQueriesARP, dbID);
+			clients[clientID].numQueriesARP = 0;
+
+			// Store hostname if available
+			if(hostname != NULL && strlen(hostname) > 0)
+			{
+				// Store host name
+				dbquery("UPDATE network "\
+				        "SET name = \"%s\" "\
+				        "WHERE id = %i;",\
+				        hostname, dbID);
+			}
+		}
+		// else:
+		// Device in database but not known to Pi-hole: No action required
+
+		// Count number of processed ARP cache entries
+		entries++;
+	}
+
+	// Actually update the database
+	dbquery("COMMIT");
+
+	// Debug logging
+	if(debug) logg("ARP table processing (%i entries) took %.1f ms", entries, timer_elapsed_msec(ARP_TIMER));
+
+	// Close file handle
+	fclose(arpfp);
+
+	// Close database connection
+	dbclose();
+}
+
+char* getMACVendor(const char* hwaddr)
+{
+	struct stat st;
+	if(stat(FTLfiles.macvendordb, &st) != 0 || strlen(hwaddr) != 17)
+	{
+		// File does not exist or MAC address is incomplete
+		if(debug) logg("getMACVenor(%s): %s does not exist or MAC invalid (length %lu)", hwaddr, FTLfiles.macvendordb, strlen(hwaddr));
+		return "";
+	}
+
+	sqlite3 *macdb;
+	int rc = sqlite3_open_v2(FTLfiles.macvendordb, &macdb, SQLITE_OPEN_READONLY, NULL);
+	if( rc ){
+		logg("getMACVendor(%s) - SQL error (%i): %s", hwaddr, rc, sqlite3_errmsg(macdb));
+		sqlite3_close(macdb);
+		return "";
+	}
+
+	char *querystr = NULL;
+	// Only keep "XX:YY:ZZ" (8 characters)
+	char * hwaddrshort = strdup(hwaddr);
+	hwaddrshort[8] = '\0';
+	rc = asprintf(&querystr, "SELECT vendor FROM macvendor WHERE mac LIKE \"%s\";", hwaddrshort);
+	if(rc < 1)
+	{
+		logg("getMACVendor(%s) - Allocation error (%i)", hwaddr, rc);
+		sqlite3_close(macdb);
+		return "";
+	}
+	free(hwaddrshort);
+
+	sqlite3_stmt* stmt;
+	rc = sqlite3_prepare_v2(macdb, querystr, -1, &stmt, NULL);
+	if( rc ){
+		logg("getMACVendor(%s) - SQL error prepare (%s, %i): %s", hwaddr, querystr, rc, sqlite3_errmsg(macdb));
+		sqlite3_close(macdb);
+		return "";
+	}
+	free(querystr);
+
+	char *vendor = NULL;
+	rc = sqlite3_step(stmt);
+	if(rc == SQLITE_ROW)
+	{
+		const unsigned char *result = sqlite3_column_text(stmt, 0);
+		// Need to use sprintf(%s) to convert unsigned char* to
+		// standard C string literals (which are char*)
+		if(asprintf(&vendor, "%s", result) < 1)
```

Comment:
How so? In C++, I'd use a `reinterpret_cast`, however, I'm not sure how one would do it in `C`. can we be sure that `(signed) char` and `unsigned char` have the same representation and alignment? Hostnames should only use the 7bit ASCII table (no extended chars), so range from 0 to 127. This would fit both into `un/signed char*`. But does this mean that they be used interchangeable here?

Reply:
As long as the data is only ASCII, the upper range of the `unsigned char` should not be used, so the cast is safe. The cast would be:
```c
const char *result = (char *) sqlite4_column_text(stmt, 0);
```